rinja_parser/src/expr.rs

```diff
@@ -190,26 +191,42 @@ impl<'a> Expr<'a> {
     expr_prec_layer!(band, shifts, token_bitand);
     expr_prec_layer!(shifts, addsub, alt((tag(">>"), tag("<<"))));
     expr_prec_layer!(addsub, muldivmod, alt((tag("+"), tag("-"))));
-    expr_prec_layer!(muldivmod, is_defined, alt((tag("*"), tag("/"), tag("%"))));
+    expr_prec_layer!(muldivmod, is_as, alt((tag("*"), tag("/"), tag("%"))));
 
-    fn is_defined(i: &'a str, level: Level) -> ParseResult<'a, WithSpan<'a, Self>> {
+    fn is_as(i: &'a str, level: Level) -> ParseResult<'a, WithSpan<'a, Self>> {
         let start = i;
         let (i, lhs) = Self::filtered(i, level)?;
-        let (i, rhs) = opt(preceded(
-            ws(keyword("is")),
-            opt(terminated(opt(keyword("not")), ws(keyword("defined")))),
-        ))(i)?;
-        let is_neg = match rhs {
-            None => return Ok((i, lhs)),
-            Some(None) => {
+        let (j, rhs) = opt(ws(identifier))(i)?;
+        let i = match rhs {
+            Some("is") => j,
+            Some("as") => {
+                let (i, target) = opt(identifier)(j)?;
+                let target = target.unwrap_or_default();
+                return match target {
+                    "bool" | "f32" | "f64" | "i8" | "i16" | "i32" | "i64" | "i128" | "isize"
+                    | "u8" | "u16" | "u32" | "u64" | "u128" | "usize" => {
+                        Ok((i, WithSpan::new(Self::As(Box::new(lhs), target), start)))
+                    }
+                    _ => Err(nom::Err::Failure(ErrorContext::new(
+                        "operator `as` expects the name of a primitive type on its right-hand side",
+                        j,
+                    ))),
+                };
+            }
+            _ => return Ok((i, lhs)),
+        };
+
+        let (i, rhs) = opt(terminated(opt(keyword("not")), ws(keyword("defined"))))(i)?;
+        let ctor = match rhs {
+            None => {
                 return Err(nom::Err::Failure(ErrorContext::new(
                     "expected `defined` or `not defined` after `is`",
                     // We use `start` to show the whole `var is` thing instead of the current token.
                     start,
                 )));
             }
-            Some(Some(None)) => false,
-            Some(Some(Some(_))) => true,
+            Some(None) => Self::IsDefined,
```

Comment:
Nice improvement! I wonder how I could not think about it directly...

Reply:
I could have thought about that before approving your PR, too. Sometimes one notices the obvious only after spending some time on other stuff. :)